File: Templates/DSU.py
class DSU:
    def __init__(self, n):
        self.par = [x for x in range(0, n + 1)]
        self.siz = [1] * (n + 1)

    def find(self, x):
        while x != self.par[x]:
            x = self.par[x] = self.par[self.par[x]]
        return x

    def same(self, x, y):
        return self.find(x) == self.find(y)

    def merge(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return False
        if self.siz[x] < self.siz[y]:
            x, y = y, x
        self.par[y] = x
        self.siz[x] += self.siz[y]
        return True

    def size(self, x):
        return self.siz[self.find(x)]

File: Templates/DSU.go
type DSU struct {
	par []int
	siz []int
}

func newDSU(n int) *DSU {
	p := make([]int, n + 1)
	s := make([]int, n + 1)
	for i := 1; i <= n; i++ {
		p[i] = i
		s[i] = 1
	}
	return &DSU{p, s}
}

func (dsu *DSU) find(x int) int {
	for x != dsu.par[x] {
		dsu.par[x] = dsu.par[dsu.par[x]]
		x = dsu.par[x]
	}
	return x
}

func (dsu *DSU) same(x int, y int) bool {
	return dsu.find(x) == dsu.find(y)
}

func (dsu *DSU) size(x int) int {
	return dsu.siz[dsu.find(x)]
}

func (dsu *DSU) merge(x int, y int) bool {
	fx := dsu.find(x)
	fy := dsu.find(y)

	if fx == fy {
		return false
	}

	if dsu.siz[fx] < dsu.siz[fy] {
		fx, fy = fy, fx
	}

	dsu.siz[fx] += dsu.siz[fy]
	dsu.par[fy] = fx
	return true
}